### ISSUE TITLE
CounterCacheBehavior with BelongsToMany Associations and `through` option not update field on delete the entity associated

### ISSUE THREAD

**Issue by ChristopherHS** [cakephp/cakephp#16356]
## What you did and what happened

```php
<?php
class RelatedsVideosTable extends BaseTable
{
    public function initialize(array $config): void
    {
        parent::initialize($config);

        $this->setTable('relateds_videos');
        $this->setDisplayField(['video_id', 'related_id']);
        $this->setPrimaryKey(['video_id', 'related_id']);

        $this->addBehavior('CounterCache', [
            'Videos' => [
                'related_count' => function (EventInterface $event, RelatedsVideo $relatedsVideo, RelatedsVideosTable $table) {
                    $query = $table->Videos->find('active')
                        ->innerJoinWith('Relateds', function (Query $query) {
                            return $query->find('active');
                        })
                        ->where([$table->Videos->aliasField('id') => $relatedsVideo->video_id]);

                    return $query->count();
                },
            ],
        ]);

        $this->belongsTo('Videos', [
            'foreignKey' => 'video_id',
            'joinType' => 'INNER',
        ]);
        $this->belongsTo('Relateds', [
            'foreignKey' => 'related_id',
            'joinType' => 'INNER',
        ]);
    }
}

class RelatedsTable extends BaseTable
{
    public function initialize(array $config): void
    {
        parent::initialize($config);

        $this->setTable('videos');
        $this->setDisplayField('title');
        $this->setPrimaryKey('id');

        $this->addBehavior('Timestamp');

        $this->belongsToMany('Videos', [
            'through' => 'RelatedsVideos',
            'dependent' => true,
            'cascadeCallbacks' => true,
            'saveStrategy' => 'replace',
        ]);
    }
}

class VideosTable extends BaseTable
{
    public function initialize(array $config): void
    {
        parent::initialize($config);

        $this->setTable('videos');
        $this->setDisplayField('title');
        $this->setPrimaryKey('id');

        $this->addBehavior('Timestamp');

        $this->belongsToMany('Relateds', [
            'through' => 'RelatedsVideos',
            'dependent' => true,
            'cascadeCallbacks' => true,
            'saveStrategy' => 'replace',
        ]);
    }
}
```
###### Table: `videos`

| id | title           | related_count | active |
|----|-----------------|---------------|--------|
| 1  | Video 1         | 2             | 1      |
| 2  | Video 2         | 0             | 1      |
| 3  | Video 3 Not Active | 0             | 0      |
| 4  | Video 4         | 0             | 1      | 

###### Table: `relateds_videos`

| video_id | related_id |
|----------|------------|
| 1        | 2          |
| 1        | 4          |

### I delete video 4
```php
$this->Videos->delete($this->Videos->get(4));
```
The video is deleted but the `related_count` column should update

## What you expected to happen

When I delete video 4 which is related to video 1, the related_count column is not updated.
It should go from 2 to 1

###### Table: `videos`

| id | title           | related_count | active |
|----|-----------------|---------------|--------|
| 1  | Video 1         | 1             | 1      |
| 2  | Video 2         | 0             | 1      |
| 3  | Video 3 Not Active | 0             | 0      |

###### Table: `relateds_videos`

| video_id | related_id |
|----------|------------|
| 1        | 2          |

### CakePHP Version

4.3.5

### PHP Version

8.1.2

**Comment by LordSimal** [comment]
Sorry for the long wait till someone answers but lets see if I can help 😄 

I don't have the exact table structure you have but for me it does indeed update the counter cache if I delete a record on the other side.
The main entry point you can check on your side is the `afterDelete` method inside the CounterCacheBehavior [here](https://github.com/cakephp/cakephp/blob/4.x/src/ORM/Behavior/CounterCacheBehavior.php#L181).

Can you check if you actually get in there or if this isn't triggered at all?

If it is triggered then next I would check if you get inside your closure you define for your `related_count` above.

If that isn't the case it seems that you don't get into either those 2 cases:
https://github.com/cakephp/cakephp/blob/4.x/src/ORM/Behavior/CounterCacheBehavior.php#L254 and
https://github.com/cakephp/cakephp/blob/4.x/src/ORM/Behavior/CounterCacheBehavior.php#L265

which in both cases is related to the `$updateConditions` and therefore the foreign keys.

**Comment by ChristopherHS** [comment]
Hi,

I just created a simplistic Github repository for testing.

Github repository:
https://github.com/ChristopherHS/CakePHPCounterCacheTest

Schema:
https://github.com/ChristopherHS/CakePHPCounterCacheTest/blob/master/config/schema/test.sql

Test:
`vendor/bin/phpunit tests/TestCase/Model/Table/VideosTableTest.php`

**Comment by LordSimal** [comment]
The fact that you use the same DB table on both sides of the belongsToMany creates this problem.

TLDR:
set this inside your `src/Model/Table/VideosTable.php`
```
        $this->belongsToMany('Relateds', [
            'foreignKey' => 'video_id',
            'targetForeignKey' => 'related_id',
            'through' => 'RelatedsVideos',
            'dependent' => true,
            'cascadeCallbacks' => true,
        ]);
```

and this inside your `src/Model/Table/RelatedsTable.php`
```
        $this->belongsToMany('Videos', [
            'foreignKey' => 'related_id',
            'targetForeignKey' => 'video_id',
            'through' => 'RelatedsVideos',
            'dependent' => true,
            'cascadeCallbacks' => true,
        ]);
```

But why you might ask?

Well, if you don't specify the foreign keys for either side of the belongsToMany then CakePHP just gueses it via the related DB table name. `$this->_targetForeignKey` will get the config value passed to the association from above
See https://github.com/cakephp/cakephp/blob/4.x/src/ORM/Association/BelongsToMany.php#L183

So in your case (not setting the foreign keys) the association just defaulted to a foreign key of `video_id` because your target table is `videos` and it inflects the foreign key upon that name.
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/9105243/171255526-9c85635f-464d-44ab-a587-c45744396205.png">
But you want to actually delete entries which have `2` as a `related_id`.

At least for me your tests are working just fine if I set the individual foreign keys on the belongsToMany association 😄

**Comment by ChristopherHS** [comment]
Hi Kevin,

Thanks for your return.
Indeed, all the tests pass now.

The videos on Youtube are Top, can't wait to see more advanced content...

Thanks again for your help.

**Comment by LordSimal** [comment]
Happy I could help 😁

**Comment by dieissonmartins** [comment]
Thanks my friend :)
